test/test_corpus.cc

```diff
@@ -601,6 +602,110 @@ TEST_P(LSPTest, All) {
+                                   make_unique<Position>(error->range->end->line, error->range->end->character)),
+                make_unique<CodeActionContext>(move(diagnostics)));
+            auto req = make_unique<RequestMessage>("2.0", nextId++, LSPMethod::TextDocumentCodeAction, move(params));
+            auto responses = lspWrapper->getLSPResponsesFor(LSPMessage(move(req)));
+            EXPECT_EQ(responses.size(), 1) << "Did not receive a response for a codeAction request.";
+
+            if (responses.size() == 1) {
+                auto &msg = responses.at(0);
+                EXPECT_TRUE(msg->isResponse());
+                if (msg->isResponse()) {
+                    auto &response = msg->asResponse();
+                    ASSERT_TRUE(response.result) << "Code action request returned error: " << msg->toJSON();
+                    auto &receivedCodeActionResponse =
+                        get<variant<JSONNullObject, vector<unique_ptr<CodeAction>>>>(*response.result);
+
+                    EXPECT_FALSE(get_if<JSONNullObject>(&receivedCodeActionResponse));
+
+                    if (!get_if<JSONNullObject>(&receivedCodeActionResponse)) {
+                        UnorderedMap<string, unique_ptr<CodeAction>> receivedCodeActionsByTitle;
+                        auto &receivedCodeActions = get<vector<unique_ptr<CodeAction>>>(receivedCodeActionResponse);
+                        for (auto &codeAction : receivedCodeActions) {
+                            bool codeActionTitleUnique =
+                                receivedCodeActionsByTitle.find(codeAction->title) == receivedCodeActionsByTitle.end();
+                            EXPECT_TRUE(codeActionTitleUnique)
+                                << "Found code action with duplicate title: " << codeAction->title;
+
+                            if (codeActionTitleUnique) {
+                                receivedCodeActionsByTitle[codeAction->title] = move(codeAction);
+                            }
+                        }
+
+                        vector<shared_ptr<ApplyCodeActionAssertion>> matchedCodeActionAssertions;
+
+                        // Check any code action assertions for the same error range, i.e. verify that when applied,
+                        // they match the expectation.
+                        auto it = applyCodeActionAssertions.begin();
+                        while (it != applyCodeActionAssertions.end()) {
+                            auto codeActionAssertion = it->get();
+                            // TODO(sushain): what if there are multiple errors on the same range?
+                            if (cmpRanges(*codeActionAssertion->range, *error->range)) {
+                                auto it2 = receivedCodeActionsByTitle.find(codeActionAssertion->title);
+                                EXPECT_NE(it2, receivedCodeActionsByTitle.end());
```

Comment:
There should be a useful error message here, e.g.:

`EXPECT_NE(..., ...) << fmt::format("Did not receive expected code action {} for error {}...");`

Reply:
You might also want to consider using `prettyPrintRangeComment` in `position_assertions.cc` to show in the error which code action is implicated (maybe define a `CodeActionAssertion::checkAll` method like I do w/ errors?).